open-craft/opencraft
# BB-2131: Implement theming customization API and frontend models

This PR introduces a new theming customization API with a custom PATCH endpoint to allow editing single fields from each `draft_theme_config`.
This also changes the behavior of some endpoints, adding a automatically generated serializer from the theme schema, so that the frontend properly parses these values.

**Setup instructions:**
1. Checkout this branch.
2. Run vagrant `vagrant up && vagrant ssh`.
3. Check if all migrations are applied and run Ocim `make run.dev`.
4. Open another terminal window and navigate to `frontend/`.
5. Set up the frontend with `npm install` and them set up the API client running `./scripts/build-api-client.sh`.

**Testing instructions:**
1. Create a new account using the new Ocim UX frontend (http://localhost:3000/registration/domain)
2. Go to http://localhost:5000/admin/registration/betatestapplication/ and check if the betatest application was correctly created, and the `Draft Theme Configuration JSON` field is filled.
3. Go to http://localhost:5000/api/swagger/ and test the `
​/v2​/instances​/openedx_config​/{id}​/theme_config​/` API endpoint.
* Change a single variable and check that it is correctly returned and saved:
```
{"footer-bg": "#fafafa"}
```
* Erase a default variable (such as `main-color`) and check that it goes back to the theme default:
```
{"main-color": ""}
```
* Set one of the optional theme variables from [here|https://github.com/open-craft/opencraft/blob/new-console/instance/schemas/theming.py#L76] and then erase it:
```
{"main-nav-color": "#fefefe"}
```
Then:
```
{"main-nav-color": ""}
```
4. Log in the new console using the account you just created (http://localhost:3000/login).
5. Click on `Theme`, then `Preview and Colors` and click on the color fields, pick a color and lose focus on the component. You should see a network request patching that exact fields. All theming fields should be greyed out while the request is being processed.

**Notes:**

1. No need to review files under `frontend/packages/api_client` as they are auto-generated by swagger codegen.

**Reviewers:**
- [ ] @alanoe
